test/mailers/concerns/workflow/mail_test.rb

```diff
@@ -99,5 +99,36 @@ class MailTest < ActionMailer::TestCase
         current_user: project.owner
       ).state_changed.deliver_now
     end
+
+    test 'transitioned_to_rejected' do
+      project = projects(:dummy_project)
+      user    = users(:application_manager_one)
+
+      project.assigned_user = user
+      project.save!(validate: false)
+
+      project.project_states.build(
+        state: workflow_states(:rejected),
+        user: user
+      ).save!(validate: false)
+
+      email = ProjectsMailer.with(
```

Comment:
should the building of the rejected state automatically trigger ProjectMailer?

Reply:
Depends on who you ask.

I'm not a fan of side effect logic in the persistence layer. It has no bearing on what is required for saving a record, it can encourage tight coupling between components, and in some cases can lead to models having more LoC around mailing and notifications than any other actual behaviour it exists for. Generally, I'd rather interested parties to act in response to an event such as a record saving (e.g. in a pub/sub paradigm), or whatever is coordinating business logic (be it a controller, job, service object, command, Rake task etc.) to orchestrate things as required.

YMMV.

Reply:
Yep I wasn't quite sure if we were expecting the state change to trigger the mail for us.